(kubectl apply): Split patching types into functions and refactorings

ardaguclu opened the pull request:
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Patch type determination is done by checking existence
of resource in schema and if it exists, it uses strategic merge patch.
Otherwise, like for CRDs, it uses merge patch type.

Currently, this code portion is not easily extensible and this PR
splits required checks into their own function to increase extensibility.

This PR is prerequisite of https://github.com/kubernetes/kubernetes/pull/110454

#### Which issue(s) this PR fixes:
Fixes #

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

ardaguclu committed:
(kubectl apply): Split patching types into functions and refactorings

Patch type determination is done by checking existence
of resource in schema and if it exists, it uses strategic merge patch.
Otherwise, like for CRDs, it uses merge patch type.

Currently, this code portion is not easily extensible and this PR
splits required checks into their own function to increase extensibility.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -111,54 +113,30 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
-				return nil, nil, cmdutil.AddSourceToErr(fmt.Sprintf(createPatchErrFormat, original, modified, current), source, err)
-			}
-			patch, err = strategicpatch.CreateThreeWayMergePatch(original, modified, current, lookupPatchMeta, p.Overwrite)
+			patch, err = p.buildStrategicMergeFromBuiltins(p.Mapping.GroupVersionKind, original, modified, current)
```
After carefully reading this https://github.com/kubernetes/kubernetes/pull/110454#issuecomment-1155451009, If I'm not missing something, this PR covers required refactorings. Only missing is that if builtin strategic merge fails, we are not falling back to merge patch as you suggested.

Since this PR is cleanup, I did not change the functionality.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -111,54 +113,30 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
+
+	patchType, err := p.getPatchType(p.Mapping.GroupVersionKind)
+	if err != nil {
+		return nil, nil, cmdutil.AddSourceToErr(fmt.Sprintf("getting patch type for %v:", p.Mapping.GroupVersionKind), source, err)
```
`source` is given to the `patchSimple` method merely so that we can build the error, which we almost always build in the same way, I think we can do better than that. The `AddSourceToErr` function isn't even really used properly.

I think you can:
1. Stop passing "source" to `patchSimple`, change the calling function to call `AddSourceToErr` with the returned error.
2. Only return mere, normal errors from this method
3. Don't pass `source` anymore to `buildMergePatch`

I suspect you could always use the `createPatchErrFormat` in the calling function, without having to hard-code it at the top of the file.

ardaguclu replied:
I removed `source` field from `patchSimple` because you are right. We are just passing it to pass `AddSourceToErr`.

Only downside of this change is we need to discriminate errors(ie. https://github.com/kubernetes/kubernetes/blob/609db7ed0b1f2839e414c17d29fe4d76edc994bd/staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go#L105 and https://github.com/kubernetes/kubernetes/blob/609db7ed0b1f2839e414c17d29fe4d76edc994bd/staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go#L111) for backwards compatibility. Thus, I started passing error verb to caller function.

apelisse replied:
"backwards compatibility", how bad is it if we change the error a little bit? Returning the two values is pretty ugly :-/
I don't think we make any guarantees that the error strings have to be constant (it may break some tests possibly, but I'm curious).

ardaguclu replied:
Maybe I can just return err(after wrapping them with errVerb using errors.Wrap...) to caller. That will change error messages but this will be much more clearer.

apelisse replied:
I think that's fine, you can ask sig-cli @seans3 @soltysh

apelisse reviewed (commented):
That's a lot simpler, but I left additional suggestions, thanks!

Note that my suggestion implies that the error messages will change a little, I think that's ok.

ardaguclu committed:
Stop passing source as parameter

This commit removes passing source field as parameter. Instead,
this commit returns error verb and error functions to caller function.
Caller function can add source field by generating correct message.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -180,6 +158,72 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 }
 
+// getPatchType returns correct PatchType for the given GVK.
+func (p *Patcher) getPatchType(gvk schema.GroupVersionKind) (types.PatchType, error) {
```
Actually not yet. After this cleanup PR is complete, this https://github.com/kubernetes/kubernetes/pull/110454 will add `p.OpenapiSchema` usage.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -180,6 +158,72 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 }
 
+// getPatchType returns correct PatchType for the given GVK.
+func (p *Patcher) getPatchType(gvk schema.GroupVersionKind) (types.PatchType, error) {
```
Yeah, I'm not super convinced yet, now you have the logic deciding what type of patch that is slightly split into two locations: This, and some of the logic is above, I think it's worse.

ardaguclu replied:
Sorry, I think I am missing something. What my plan is after two PR are merged, this function will look like this https://github.com/kubernetes/kubernetes/pull/110454/files#diff-5dbe477b3463f52cdf1d6913df2f8f3e03907f37306d90e85d4e5386ca6ee0e6R164-R186

apelisse replied:
yeah, for now I don't really mind what it will look like in the future :-) I just don't think this looks great because the logic is split, IMHO, in a weird way.

ardaguclu replied:
Fair enough:). I'll move this logic back to main function

apelisse replied:
Thanks, maybe I won't like it lol sorry about that :-)

ardaguclu committed:
Move getPatchType logic back to main function

ardaguclu committed:
Use error wrapping in patchSimple

ardaguclu committed:
Update dependencies for kubectl

ardaguclu commented:
@apelisse could you PTAL?, thanks.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -172,34 +158,89 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
+// buildStrategicMergeFromOpenAPI builds patch from OpenAPI if it is enabled.
+// This is used for core types which is published in openapi.
+func (p *Patcher) buildStrategicMergeFromOpenAPI(gvk schema.GroupVersionKind, original, modified, current []byte) ([]byte, error) {
+	if s := p.OpenapiSchema.LookupResource(gvk); s != nil {
```
I'd possibly remove this from the function. Returning `nil, nil` is not great.

apelisse replied:
Since Go doesn't let you check if p.OpenapiSchema and assign at the same time, I'll suggest we add a function to the patcher like the following:
```golang
func (p *Patcher) findOpenAPIResource(gvk schema.GroupVersionKind) Resource {
    if p.OpenapiSchema == nil {
        return nil
    }
    return p.OpenapiSchema.LookupResource(gvk)
}
```

You can then use it above L138.

ardaguclu replied:
@apelisse done, PTAL?

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:
```diff
@@ -98,69 +102,51 @@ func (p *Patcher) delete(namespace, name string) error {
-	case runtime.IsNotRegisteredError(err):
-		// fall back to generic JSON merge patch
+
+	patchType := types.StrategicMergePatchType
```
I think you should remove this variable, it's not useful.

ardaguclu replied:
It is used in switch https://github.com/kubernetes/kubernetes/pull/110619/files#diff-5dbe477b3463f52cdf1d6913df2f8f3e03907f37306d90e85d4e5386ca6ee0e6R129

ardaguclu committed:
Move resource lookup into its own function

ardaguclu commented:
@apelisse  As we discussed on Slack. Returning `patch` in here https://github.com/kubernetes/kubernetes/blob/383b5f676670d99dc1fba4254ba7f6a81a052ba3/staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go#L180 is redundant. It is only used in here https://github.com/kubernetes/kubernetes/blob/8fb423bfabe0d53934cc94c154c7da2dc3ce1332/staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go#L623-L630

I think, we can use `reflect.DeepEqual` instead this and remove it?. If it sounds reasonable, I'll remove that in followup PR.

apelisse commented:
I think this is a net positive right now, even though I suspect we can go a lot further.
/lgtm
/approve
